source/Hardpoint.cpp

```diff
@@ -201,36 +183,53 @@ void Hardpoint::Step()
 	// If the full reload time is elapsed, reset the burst counter.
 	if(reload <= 0.)
 		burstCount = outfit->BurstCount();
+	// If the weapon recently fired then increment any sustained fire counters.
+	if(reload > 0.)
+	{
+		if(bloomCount < outfit->BloomTime())
+			++bloomCount;
+		if(spinupCount < outfit->SpinupTime())
+			++spinupCount;
+	}
+	// Only decrement sustained fire counters if this weapon could have fired
+	// but didn't.
+	else if(!wasFiring)
+	{
+		if(bloomCount > 0.)
+			--bloomCount;
+		if(spinupCount > 0.)
+			--spinupCount;
+	}
+
 	if(burstReload > 0.)
 		--burstReload;
 	// If the burst reload time has elapsed, this weapon will not count as firing
 	// continuously if it is not fired this frame.
-	if(burstReload <= 0. && reload <= 0)
+	if(burstReload <= 0.)
 		isFiring = false;
-	if(outfit->BurstCount() > 1 && burstCount == 0)
+	// If a burst has run out then reset the burst sustained fire counters.
+	if(burstCount == 0)
 	{
-		burstSpinupCount = outfit->BurstSpinupTime();
-		if(wasFiring && spinupCount > 0)
-			--spinupCount;
-		else if(!wasFiring && spinupCount < outfit->SpinupTime())
-			++spinupCount;
+		burstBloomCount = 0.;
+		burstSpinupCount = 0.;
 	}
-	else
+	// If the weapon recently fired a burst then increment any burst
+	// sustained fire counters.
+	else if(burstReload > 0.)
 	{
-		if(wasFiring)
-		{
-			if(spinupCount > 0)
-				--spinupCount;
-			if(burstSpinupCount > 0)
-				--burstSpinupCount;
-		}
-		else
-		{
-			if(spinupCount < outfit->SpinupTime())
-				++spinupCount;
-			if(burstSpinupCount < outfit->BurstSpinupTime())
-				++burstSpinupCount;
-		}
+		if(burstBloomCount < outfit->BurstBloomTime())
+			++burstBloomCount;
+		if(burstSpinupCount < outfit->BurstSpinupTime())
+			++burstSpinupCount;
+	}
+	// Only decrement burst sustained fire counters if this weapon could
+	// have fired but didn't.
+	else if(!wasFiring)
+	{
+		if(burstBloomCount > 0.)
+			--burstBloomCount;
+		if(burstSpinupCount > 0.)
+			--burstSpinupCount;
```

Comment:
There are a couple duplicated checks in the current logic.  This should clean it up while retaining the same logic.
```suggestion
	if(reload > 0.)
	{
		// Since the weapon is firing as often as it can, increment overall sustained fire counters.
		if(bloomCount < outfit->BloomTime())
			++bloomCount;
		if(spinupCount < outfit->SpinupTime())
			++spinupCount;
	}
	else
	{
		// Since the full reload time has elapsed, reset the burst counter.
		burstCount = outfit->BurstCount();

		if(!wasFiring)
		{
			// This weapon could have fired (if burst type it could have continued to try),
			// so decrement overall sustained fire counters.
			if(bloomCount > 0.)
				--bloomCount;
			if(spinupCount > 0.)
				--spinupCount;
		}
	}

	if(burstReload > 0.)
		--burstReload;
	else
		// Set up a break in continuous firing that can be reverted by firing this frame.
		isFiring = false;
	if(burstCount == 0)
	{
		// A burst has run out so reset burst sustained fire counters.
		burstBloomCount = 0.;
		burstSpinupCount = 0.;
	}
	else if(burstReload > 0.)
	{
		// This weapon is firing as often as it can within a burst,
		// so increment burst sustained fire counters.
		if(burstBloomCount < outfit->BurstBloomTime())
			++burstBloomCount;
		if(burstSpinupCount < outfit->BurstSpinupTime())
			++burstSpinupCount;
	}
	else
	{
		// This weapon could have fired but didn't, so decrement burst sustained fire counters.
		if(!wasFiring)
		{
			if(burstBloomCount > 0.)
				--burstBloomCount;
			if(burstSpinupCount > 0.)
				--burstSpinupCount;
		}
```